test/channels.js

```diff
@@ -0,0 +1,121 @@
+const expect = Code.expect;
+
+describe('DiagnosticChannel', () => {
+
+    describe('onServerChannel', () => {
+
+        const channel = DC.channel('hapi.onServer');
+
+        it('server should be exposed on creation through the channel hapi.onServer', async () => {
+
+            let exposedServer;
+            let server;
+
+            await new Promise((resolve) => {
+
+                channel.subscribe((srv) => {
+
+                    exposedServer = srv;
+                    resolve();
+                });
+
+                server = Hapi.server();
+            });
+
+            expect(exposedServer).to.equal(server);
+        });
+    });
+
+    describe('onRouteChannel', () => {
+
+        const channel = DC.channel('hapi.onRoute');
+
+        it('route should be exposed on creation through the channel hapi.onRoute', async () => {
+
+            const server = Hapi.server();
+            let route;
+
+            await new Promise((resolve) => {
+
+                channel.subscribe((rte) => {
+
+                    route = rte;
+                    resolve();
+                });
+
+                server.route({
+                    method: 'GET',
+                    path: '/',
+                    options: { app: { x: 'o' } },
+                    handler: () => 'ok'
+                });
+            });
+
+            expect(route).to.be.an.object();
+            expect(route.settings.app.x).to.equal('o');
+        });
+    });
+
+    describe('onResponseChannel', () => {
+
+        const channel = DC.channel('hapi.onResponse');
+
+        it('response should be exposed on creation through the channel hapi.onResponse', async () => {
+
+            const server = Hapi.server();
+            let responseExposed;
+
+            server.route({ method: 'GET', path: '/', handler: () => 'ok' });
+
+            const eventPromise = new Promise((resolve) => {
+
+                channel.subscribe((res) => {
+
+                    responseExposed = res;
+                    resolve();
+                });
+            });
+
+            const response = await server.inject('/');
+            await eventPromise;
+
+            expect(response.request.response).to.equal(responseExposed);
+        });
+    });
+
+    describe('onRequestChannel', () => {
+
+        const channel = DC.channel('hapi.onRequest');
+
+        it('request should be exposed on creation through the channel hapi.onRequest', async () => {
+
+            const server = Hapi.server();
+            let requestExposed;
+
+            server.route({ method: 'GET', path: '/', handler: () => 'ok' });
+
+            const eventPromise = new Promise((resolve) => {
+
+                channel.subscribe((req) => {
+
+                    requestExposed = req;
+                    resolve();
+                });
+            });
+
+            const response = await server.inject('/');
+            await eventPromise;
+            expect(response.request).to.equal(requestExposed);
```

Comment:
Maybe add a check to verify that it has not been routed when it is triggered? Eg.

```js
             const event = new Promise((resolve) => {

                 channel.subscribe(resolve);
             });

             const req = server.inject('/');
             const request = await event;
             expect(request).to.exist();
             expect(request.route).to.not.exist();

             const response = await req;
             expect(response.request).to.shallow.equal(request);
```

Reply:
From what I found, `request` are created with the default `notFound` route :

https://github.com/hapijs/hapi/blob/c6d881ff10e3e102c2af99f117901740ee048d7d/lib/request.js#L57

Should I test that instead ?

Reply:
This sounds like a minor hapi bug – the docs state that [it is unassigned](https://github.com/hapijs/hapi/blob/c6d881ff10e3e102c2af99f117901740ee048d7d/API.md?plain=1#L3622) when `onRequest` is processed. I have prepared [a PR to fix this](#4433).

In the meantime you can test for `request.params` to equal `null` instead.